File: Roadmap/04 - CADENAS DE CARACTERES/javascript/JuSeRDev.js
let saludo = "Hola!"
let lenguaje = "JavaScript"

let saludar = saludo + " "+ lenguaje
console.log(saludar);

console.log(saludar.length)

console.log(saludar[0])
console.log(saludar[1])
console.log(saludar[2])
console.log(saludar[3])
console.log(saludar[4])
console.log(saludar[5])
console.log(saludar[6])
console.log(saludar[7])
console.log(saludar[8])
console.log(saludar[9])
console.log(saludar[10])
console.log(saludar[11])
console.log(saludar[12])
console.log(saludar[13])
console.log(saludar[14])
console.log(saludar[15])


let extraccion = saludar.substring(0,10)
console.log(extraccion)

let extraccion2 = saludar.slice(-16,-5)
console.log(extraccion2);





// const confirmacion = (palabra, palabra2)=>{
//     let invertido = ""
//     let invertido2 = ""
//     for (let i =  palabra.length -1; i >= 0 ; i--) {
//         invertido += palabra[i]
//     }
//     for (let i =  palabra2.length -1; i >= 0 ; i--) {
//         invertido2 += palabra2[i]
//     }
//     if (invertido === palabra) {
//         console.log(`la palabra: ${palabra} es un Palindromo`)
//     } else {
//         console.log(`la palabra: ${palabra} no es un Palindromo`)
//     }
//     if (invertido2 === palabra2) {
//         console.log(`la palabra: ${palabra2} es un Palindromo`)
//     } else {
//         console.log(`la palabra: ${palabra2} no es un Palindromo`)
//     }
// }

// confirmacion("casa", "lateleletal")

// const confirmacion = (palabra1, palabra2)=>{
//      //seprala la palabra en caracteres convirtiendolo en un array, luego lo pone alrevez y luego lo vuelve a unir en una cadena de texto
//     let invertido1 = palabra1.split("").reverse().join("")
//     let invertido2 = palabra2.split("").reverse().join("")
//     if (palabra1 === invertido1) {
//         console.log(`la plabra ${palabra1} es un Palindromo`)
//     } else {
//         console.log(`La palabra ${palabra1} no es un Palindromo`)
//     }
//     if (palabra2 === invertido2) {
//         console.log(`la plabra ${palabra2} es un Palindromo`)
//     } else {
//         console.log(`La palabra ${palabra2} no es un Palindromo`)
//     }
    
// }
// confirmacion("casa", "oso")



const invertirPalabra = (palabra)=> palabra.split("").reverse().join("")
const esPalindromo = (palabra) =>{
    const invertida = invertirPalabra(palabra)
    if (invertida === palabra) {
        return `la palabra: "${palabra}" es un palindromo`
    } else{
        return `la palabra "${palabra}" no es un palindromo`
    }
}

const esAnagrama = (opcion1, opcion2)=>{
    let separacion1 = opcion1.split("").sort()
    let separacion2 = opcion2.split("").sort()
    console.log(separacion1)
    console.log(separacion2)
    if (separacion1.length === separacion2.length && separacion1.join("") === separacion2.join("")) {
        return `la palabras 1: ${opcion1} y 2: ${opcion2} son un anagrama`
    } else{
        return `la palabras 1: "${opcion1}" y 2: "${opcion2}" no son un anagrama`
    }
}

const esIsograma = (palabra1, palabra2)=>{
    let isograma = new Set();
    (palabra1 + palabra2).split("").forEach(letra => isograma.add(letra))
    if (isograma.size === (palabra1.length + palabra2.length)) {
        console.log("si");
        return `Las palabras "${palabra1}" y "${palabra2}" son isogramas`
    } else {
        console.log("no");
        
        return `Las palabras "${palabra1}" y "${palabra2}" no son isogramas`
    }
}

const confirmacion = (palabra1, palabra2)=>{
    let resultado = []
    
    resultado.push(esPalindromo(palabra1))
    resultado.push(esPalindromo(palabra2))
    resultado.push(esAnagrama(palabra1, palabra2))
    resultado.push(esIsograma(palabra1, palabra2))
    
    console.log(esPalindromo(palabra1));
    console.log(esPalindromo(palabra2));
    console.log(esAnagrama(palabra1, palabra2))
    console.log(esIsograma(palabra1, palabra2))
//<<<<<<< main
    
=======

//>>>>>>> main
    return resultado.join("\n")
    
}

console.log(confirmacion("casa", "oso"))

//<<<<<<< main
//=======
console.log(confirmacion("perro", "lateleletal"))
//>>>>>>> main

File: Roadmap/04 - CADENAS DE CARACTERES/python/juserdev.py
""" /*
 * EJERCICIO:
 * Muestra ejemplos de todas las operaciones que puedes realizar con cadenas de caracteres
 * en tu lenguaje. Algunas de esas operaciones podrían ser (busca todas las que puedas):
 * - Acceso a caracteres específicos, subcadenas, longitud, concatenación, repetición, recorrido,
 *   conversión a mayúsculas y minúsculas, reemplazo, división, unión, interpolación, verificación...
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea un programa que analice dos palabras diferentes y realice comprobaciones
 * para descubrir si son:
 * - Palíndromos
 * - Anagramas
 * - Isogramas
 */ """

saludo = "hola"
lenguaje = "python"

#subcadena
print(saludo[0:3])
print(saludo[1:3])

#acceso caracteres especificos
print(saludo[0])

# logitud
print(len(saludo))

#concatenacion
print(saludo + " " + lenguaje)

#multiplicacion
print(saludo * 3)

#reccorido
for letra in saludo:
  print(letra)

for letra in lenguaje:
  print(letra)


#conversion a mayusculas
saludar = saludo + " " + lenguaje
saludar_mayusculas = saludar.upper()
print(saludar_mayusculas)

#convertir a minusculas
saludar_minusculas = saludar_mayusculas.lower()
print(saludar_minusculas)

#inicial mayuscula
print(saludar.capitalize())
print(saludar.title(), "este es ima ´riena de totñe()") # convierte la primera letra de cada palabra en mayuscula

# reemplazo
reemplazo = saludo.replace("h", "J")
print(reemplazo)

# divisio
division = saludar.split(" ")
print(division)

# union
# -> es una manera un poco burda de unir palabras pero funciona
oracion = " ".join(division)
print(oracion)

# Interporacion
Saludo_interpolado = f"{saludo} {lenguaje}"
print(Saludo_interpolado)

# verificacion
print("a" in lenguaje)
print("y" in lenguaje)

# busqueda
print(saludar.find("python"))

# transformacion en lista de caracteres
print(list(saludar))

# verificacion

print(saludar.isnumeric())
print(saludar.isalnum())
print(saludar.isalpha())

### dificulatad extra

def analisis(palabra1: str, palabra2: str):
  #palindromo
  print(f"{palabra1} es un palindromo --> {palabra1 == palabra1[::-1]}")
  print(f"{palabra2} es un palindromo --> {palabra2 == palabra2[::-1]}")

  #anagrama
  print(f"{palabra1} es un anagrama de {palabra2}? {sorted(palabra1) == sorted(palabra2)}")
  
  #isograma

  def isograma(palabra: str) -> bool:
    palabra_dic = dict()

    for letra in palabra:
      palabra_dic[letra] = palabra_dic.get(letra, 0) + 1

    isograma = True
    values = list(palabra_dic.values())
    isograma_len = values[0]

    for contar_palabras in values:
      if contar_palabras != isograma_len:
        isograma = False
        break
    return isograma

  print(f"{palabra1} es un isograma= -> {isograma(palabra1)}")
  print(f"{palabra1} es un isograma= -> {isograma(palabra2)}")
  
  isograma(palabra1)
  isograma(palabra2)
  
analisis("radar", "anilina")
analisis("cara", "arca")